### PR TITLE
[3.3] Backport CI config from master

### PR DESCRIPTION
**Pull request by vstinner**

### CODE REVIEW

**Comment by vstinner** [comment]
This PR is more or less a copy of the PR #2475 adapted for the 3.3 branch. Right now, I didn't backport the test_os change, since I got conflict and I'm not sure that the test_os os.walk() test also fails on 3.3.

**Comment by vstinner** [comment]
The Travis CI job failed with:

```
$ if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then ./python Tools/scripts/patchcheck.py --travis $TRAVIS_PULL_REQUEST; fi

Getting the list of files that have been added/changed ... need a checkout to get modified files
```

The AppVeyor job is full of errors, it seems like the build system is unable to open the Python 3.3 PCbuild files.

**Comment by vstinner** [comment]
@zooba, @zware, @brettcannon: Which Visual Studio is needed to build Python 3.3 on Windows? Any idea how to adapt AppVeyor configuration to fix it?

**Comment by vstinner** [comment]
If it's just impossible to build Python 3.3 on AppVeyor, hum, too bad. We might remove the AppVeyor CI in that case. If at least we could have Travis CI for Linux, it would already be a first enhancement!

**Comment by zooba** [comment]
No idea, never tried to build it. I'd assume it's the same configuration as 3.4.

Looking at the build log, it probably needs to be pointed specifically at the VS 2010 MSBuild.exe, rather than any newer version.

**Comment by vstinner** [comment]
I added 2 more commits:
* Travis CI: remove patchcheck.py test -- try to fix the Travis CI/Linux
* replace 512 bit dh key with a 2014 bit one (closes #23844): backport a7eaf56a6d1de20a7a36fde03b3512fa5f9f395d to fix test_dh_params() of test_ssl, needed for recent OpenSSL

test_protocol_sslv3() and test_options() of test_ssl still fail because of SSLv3. OpenSSL now disables SSLv3 by default: we should backport a9dcdabccb1a1f7c76030c0b188ecaf7ab599e57 from the 3.4 branch, but I would prefer to do that in a separated commit since this change not only changes tests but also Modules/_ssl.c.

**Comment by vstinner** [comment]
Ok, I removed the [WIP] tag: I consider that this PR is now ready for a review.

Summary of the PR for Python 3.3:

* Add CI configuration to get 3 CI: Travis CI Linux, Travis CI macOS and AppVeyor CI
* I adapted and simplifiy the CI configuration to 3.3
* I backported two test fixes: "replace 512 bit dh key with a 2014 bit one" to fix test_ssl, "bpo-30231: Remove skipped test_imaplib tests" to fix test_imaplib. These change only change Lib/test/

CI status:

* Travis CI Linux: test_site fails
* Travis CI macOS: 5 tests fail, test__osx_support test_distutils test_site test_socket test_sysconfig -- sorry, I didn't look at these ones. I prefer to focus on Linux.
* AppVeyor: broken, the compilation fails badly. I don't know how to debug/fix that. Maybe we should just remove this CI?

**Comment by zware** [comment]
I think getting AppVeyor working with 3.3 is probably more effort than it is worth.  Various rewrites of the batch files in PCbuild (build.bat, rt.bat, etc.) were not backported that far, and probably don't need to be.

FTR, 3.3 was built with VS 2010.

**Comment by ned-deily** [comment]
I agree with not trying to get AppVeyor working.  3.3 is just about to be end-of-lifed.  Basically, we just need to get the existing small backlog of security patches and PRs updated and merged and then that will likely be it for 3.3, assuming no new security issues arise between now and 2017-09-29.  In any case, let's hold off on making a decision whether to merge this until we've had a chance to review the 3.3 backlog.

File: .github/appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+version: 3.3.6+.{build}
```
Comment (ned-deily) [inline .github/appveyor.yml]: As noted in the comments above, let's not try to deal with appveyor.  There should only be a handful of further commits to the 3.3 branch before it is EOLed.  So delete this file.

File: .github/CONTRIBUTING.rst [hunk 2]
```diff
@@ -0,0 +1,47 @@
+Contributing to Python
```
Comment (ned-deily) [inline .github/CONTRIBUTING.rst]: There's no reason to add this file for 3.3.  Just delete it, please.

**Review by ned-deily — CHANGES_REQUESTED** [review]

File: .github/CONTRIBUTING.rst (hunk 2 above)
Comment (vstinner) [reply .github/CONTRIBUTING.rst]: deleted

File: .github/appveyor.yml (hunk 1 above)
Comment (vstinner) [reply .github/appveyor.yml]: I deleted AppVeyor.

**Comment by vstinner** [comment]
Oh, Travis CI success on Linux: "All checks have passed"

Strange. On my Fedora 25, test_ssl and test_ftplib fail (maybe an error related to SSLv3, I don't know exactly). Well, it's a good thing, the CI pass :-)

**Comment by vstinner** [comment]
The macOS CI has 4 failing tests: test__osx_support test_distutils test_socket test_sysconfig. Examples:

@ned-deily: Should I try to fix these failures, remove the macOS CI from Travis config, or just do nothing?

```
======================================================================

FAIL: test_get_platform_osx (test.test__osx_support.Test_OSXSupport)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/Users/travis/build/python/cpython/Lib/test/test__osx_support.py", line 272, in test_get_platform_osx

    self.assertEqual(('macosx', '10.6', 'fat'), result)

AssertionError: Tuples differ: ('macosx', '10.6', 'fat') != ('macosx', '10.6', ' ')

test_deployment_target_default (distutils.tests.test_build_ext.BuildExtTestCase) ... /var/folders/my/m6ynh3bn6tq06h7xr3js0z7r0000gn/T/tmpb6az8m/deptargetmodule.c:7:2: error: "Unexpected target"

#error "Unexpected target"

 ^

======================================================================

ERROR: testFDPassEmpty (test.test_socket.RecvmsgSCMRightsStreamTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/Users/travis/build/python/cpython/Lib/test/test_socket.py", line 2620, in testFDPassEmpty

    len(MSG), 10240),

  File "/Users/travis/build/python/cpython/Lib/test/test_socket.py", line 1728, in doRecvmsg

    result = sock.recvmsg(bufsize, *args)

OSError: [Errno 12] Cannot allocate memory


======================================================================

FAIL: test_get_platform (test.test_sysconfig.TestSysConfig)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/Users/travis/build/python/cpython/Lib/test/test_sysconfig.py", line 179, in test_get_platform

    self.assertEqual(get_platform(), 'macosx-10.4-fat')

AssertionError: 'macosx-10.4-x86_64' != 'macosx-10.4-fat'

- macosx-10.4-x86_64

+ macosx-10.4-fat
```

**Comment by vstinner** [comment]
Summary of the PR:

* Add  .travis.yml to run the Python test suite on Linux and macOS
* Remove 2 tests from test_imaplib using a public IMAP server which doesn't accept our private x509 client certificate. These tests were already removed in 2.7, 3.5, 3.6 and master.
* Python test suite pass on Travis CI Linux
* 4 tests fails on Travis CI macOS

AppVeyor and CodeCov are not part of this PR, I consider that we don't need them for Python 3.3. For AppVeyor, it's just too hard to setup the CI and it was decided that it's not worth it (read comments above).

IMHO a working CI on Linux is already a huge improvement compared to the current state: no more CI at all. We removed the Python 3.3 buildbots long time ago...

```
$ git diff 3.3.. --stat
 .travis.yml              | 63 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 Lib/test/test_imaplib.py | 10 ----------
 2 files changed, 63 insertions(+), 10 deletions(-)
```

**Review by brettcannon — APPROVED** [review]

**Comment by ned-deily** [comment]
"The macOS CI has 4 failing tests."  You should remove the macOS CI configuration.  Vanilla 3.3.x has  never functioned properly on macOS versions beyond OS X 9.  3.3.x was already in security-fix mode when OS X 10.10 was released which broke various things.  It would be out-of-scope to try to fix things now as there has been no demand for it that I'm aware of and it would be sending the wrong message now; in any case, some downstream distributors have backported 3.3 support long ago if anyone really needs it.

**Comment by vstinner** [comment]
> You should remove the macOS CI configuration.

Ok, done.

**Review by ned-deily — APPROVED** [review]

**Comment by vstinner** [comment]
Yeah! A CI for 3.3, that's so cool ;-)

**Pull request merged**